src/tracing.ts

```diff
@@ -0,0 +1,173 @@
+import * as PluginTypes from './plugin-types';
+import {TraceAgent} from './trace-api';
+import {pluginLoader, PluginLoaderConfig} from './trace-plugin-loader';
+import {traceWriter, TraceWriterConfig} from './trace-writer';
+import {Component, FORCE_NEW, Forceable, packageNameFromPath, Singleton} from './util';
+
+export interface TopLevelConfig {
+  enabled: boolean;
+  logLevel: number;
+  clsMechanism: CLSMechanism;
+}
+
+// PluginLoaderConfig extends TraceAgentConfig
+export type NormalizedConfig =
+    (TraceWriterConfig&PluginLoaderConfig&TopLevelConfig)|{enabled: false};
+
+/**
+ * A class that represents automatic tracing.
+ */
+export class Tracing implements Component {
+  /** An object representing the custom span API. */
+  private readonly traceAgent: TraceAgent;
+  /** A logger. */
+  private readonly logger: common.Logger;
+  /** The configuration object for this instance. */
+  private readonly config: Forceable<NormalizedConfig>;
+
+  /**
+   * Constructs a new Tracing instance.
+   * @param config The configuration for this instance.
+   */
+  constructor(config: NormalizedConfig, traceAgent: TraceAgent) {
+    this.config = config;
+    this.traceAgent = traceAgent;
+    let logLevel = config.enabled ? config.logLevel : 0;
+    // Clamp the logger level.
+    if (logLevel < 0) {
+      logLevel = 0;
+    } else if (logLevel >= common.logger.LEVELS.length) {
+      logLevel = common.logger.LEVELS.length - 1;
+    }
+    this.logger = common.logger({
+      level: common.logger.LEVELS[logLevel],
+      tag: '@google-cloud/trace-agent'
+    });
+  }
+
+
+
+  /**
+   * Logs an error message detailing the list of modules that were loaded before
+   * the Trace Agent. Loading these modules before the Trace Agent may prevent
+   * us from monkeypatching those modules for automatic tracing.
+   * @param filesLoadedBeforeTrace The list of files that were loaded using
+   * require() before the Stackdriver Trace Agent was required.
+   */
+  logModulesLoadedBeforeTrace(filesLoadedBeforeTrace: string[]) {
+    const modulesLoadedBeforeTrace: string[] = [];
+    const traceModuleName = path.join('@google-cloud', 'trace-agent');
+    for (let i = 0; i < filesLoadedBeforeTrace.length; i++) {
+      const moduleName = packageNameFromPath(filesLoadedBeforeTrace[i]);
+      if (moduleName && moduleName !== traceModuleName &&
+          modulesLoadedBeforeTrace.indexOf(moduleName) === -1) {
+        modulesLoadedBeforeTrace.push(moduleName);
+      }
+    }
+    if (modulesLoadedBeforeTrace.length > 0) {
+      this.logger.error(
+          'TraceAgent#start: Tracing might not work as the following modules',
+          'were loaded before the trace agent was initialized:',
+          `[${modulesLoadedBeforeTrace.sort().join(', ')}]`);
+    }
+  }
+
+  /**
+   * Enables automatic tracing support and the custom span API.
+   */
+  enable(): void {
+    if (this.traceAgent.isActive()) {
+      // For unit tests only.
```

Comment:
How do we make sure this path is not reached during production? Ideally code like this should be guarded with a test that can only be true during testing. It is not immediately obvious to me why `this.traceAgent.isActive()` indicates that we are running inside a unit test.

Reply:
Actually, this entire conditional isn't needed anymore. In `index.ts`, it can be statically determined that `tracing.enable()` is never called twice on the same instance.